sc-memory/sc-memory/sc_agent.tpp

```diff
@@ -0,0 +1,580 @@
+            << "` has not initiation condition and result. Check that abstract sc-agent has specified "
+               "relation `nrel_initiation_condition_and_result`.");
+
+  return m_memoryCtx.GetEdgeSource(it5->Get(2));
+}
+
+// LCOV_EXCL_START
+template <class TScEvent, class TScContext>
+ScTemplate ScAgentAbstract<TScEvent, TScContext>::GetInitiationConditionTemplate() const
+{
+  return ScTemplate();
+}
+
+// LCOV_EXCL_STOP
+
+// LCOV_EXCL_START
+template <class TScEvent, class TScContext>
+ScResult ScAgentAbstract<TScEvent, TScContext>::DoProgram(TScEvent const & event, ScAction & action)
+{
+  return ScResult(SC_RESULT_OK);
+}
+
+template <class TScEvent, class TScContext>
+ScResult ScAgentAbstract<TScEvent, TScContext>::DoProgram(ScAction & action)
+{
+  return ScResult(SC_RESULT_OK);
+}
+
+// LCOV_EXCL_STOP
+
+// LCOV_EXCL_START
+template <class TScEvent, class TScContext>
+sc_bool ScAgentAbstract<TScEvent, TScContext>::CheckResultCondition(TScEvent const &, ScAction &)
+{
+  return SC_TRUE;
+}
+
+// LCOV_EXCL_STOP
+
+template <class TScEvent, class TScContext>
+ScAddr ScAgentAbstract<TScEvent, TScContext>::GetResultCondition() const
+{
+  ScIterator5Ptr const it5 = m_memoryCtx.Iterator5(
+      GetAbstractAgent(),
+      ScType::EdgeDCommonConst,
+      ScType::EdgeDCommonConst,
+      ScType::EdgeAccessConstPosPerm,
+      ScKeynodes::nrel_initiation_condition_and_result);
+  if (!it5->Next())
+    SC_THROW_EXCEPTION(
+        utils::ExceptionInvalidState,
+        "Abstract sc-agent `"
+            << this->GetName()
+            << "` has not initiation condition and result. Check that abstract sc-agent has specified "
+               "relation `nrel_initiation_condition_and_result`.");
+
+  return m_memoryCtx.GetEdgeTarget(it5->Get(2));
+}
+
+// LCOV_EXCL_START
+template <class TScEvent, class TScContext>
+ScTemplate ScAgentAbstract<TScEvent, TScContext>::GetResultConditionTemplate() const
+{
+  return ScTemplate();
+}
+
+// LCOV_EXCL_STOP
+
+template <class TScEvent, class TScContext>
+void ScAgentAbstract<TScEvent, TScContext>::SetInitiator(ScAddr const & userAddr)
+{
+  m_memoryCtx = ScAgentContext(userAddr);
+}
+
+template <class TScEvent, class TScContext>
+void ScAgentAbstract<TScEvent, TScContext>::SetImplementation(ScAddr const & agentImplementationAddr)
+{
+  m_agentImplementationAddr = agentImplementationAddr;
+}
+
+template <class TScEvent, class TScContext>
+sc_bool ScAgentAbstract<TScEvent, TScContext>::MayBeSpecified() const
+{
+  return m_agentImplementationAddr.IsValid();
+}
+
+template <class TScEvent, class TScContext>
+std::function<void(TScEvent const &)> ScAgentAbstract<TScEvent, TScContext>::GetCallback(ScAddr const &)
+{
+  return {};
+}
+
+template <class TScEvent, class TScContext>
+ScAgent<TScEvent, TScContext>::ScAgent()
+  : ScAgentAbstract<TScEvent>(){};
+
+template <class TScEvent, class TScContext>
+template <class TScAgent, class... TScAddr>
+void ScAgent<TScEvent, TScContext>::Subscribe(
```

Comment:
Maybe you were confused by this code

 ```
 ScAddrVector subscriptionVector{subscriptionAddrs...};
  // Check that user specify agent implementation only and find subscription sc-element in knowledge base.
  if (agent.MayBeSpecified() && subscriptionVector.empty())
    subscriptionVector.emplace_back(agent.GetEventSubscriptionElement());
```
    
If agent specified in knowledge base, then `subscriptionAddrs` will be empty and `subscriptionVector` will contain only subscription sc-element from knowledge base, otherwise (when agent is not specified in knowledge base)  `subscriptionAddrs` will be not empty and `subscriptionVector` will equal to `subscriptionAddrs`. This logic is limited to the interface, you can't use this method directly for any agent class. You must use LoadAndSubscribeAgent` (for case with kb), `SubscribeAgent` (for case without kb) correspondingly. I probably should have split this method into two, but I didn't, I didn't want to duplicate the code.

Reply:
is there a test that checks that 
```
If agent specified in knowledge base, then subscriptionAddrs will be empty and subscriptionVector will contain only subscription sc-element from knowledge base, otherwise (when agent is not specified in knowledge base) subscriptionAddrs will be not empty and subscriptionVector will equal to subscriptionAddrs. 
```
?

Reply:
I'll check it later